wallet/chainntfns.go

```diff
@@ -47,42 +52,45 @@ func (w *Wallet) handleChainNotifications() {
 		// if it doesn't match the original hash returned by
 		// the notification, to roll back and restart the
 		// rescan.
-		log.Infof("Catching up block hashes to height %d, this"+
-			" might take a while", height)
-		err := walletdb.Update(w.db, func(tx walletdb.ReadWriteTx) error {
-			ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
-
-			startBlock := w.Manager.SyncedTo()
+		log.Infof("Catching up block hashes to height %d, this might "+
+			"take a while", height)
+
+		heightRange := &scanRange{start: startHeight, end: height}
+		var blockStamps []*waddrmgr.BlockStamp
+		err := scanChain(chainClient, heightRange, func(height int32,
+			hash *chainhash.Hash, header *wire.BlockHeader) error {
+
+			blockStamps = append(blockStamps, &waddrmgr.BlockStamp{
+				Hash:      *hash,
+				Height:    height,
+				Timestamp: header.Timestamp,
+			})
+			return nil
 
-			for i := startBlock.Height + 1; i <= height; i++ {
-				hash, err := client.GetBlockHash(int64(i))
-				if err != nil {
-					return err
-				}
-				header, err := chainClient.GetBlockHeader(hash)
-				if err != nil {
-					return err
-				}
+		})
+		if err != nil {
+			log.Errorf("Unable to catch up block hashes to height "+
+				"%d: %v", height, err)
+			return
+		}
 
-				bs := waddrmgr.BlockStamp{
-					Height:    i,
-					Hash:      *hash,
-					Timestamp: header.Timestamp,
-				}
-				err = w.Manager.SetSyncedTo(ns, &bs)
+		err = walletdb.Update(w.db, func(tx walletdb.ReadWriteTx) error {
+			ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+			for _, blockStamp := range blockStamps {
```

Comment:
I know that for certain backends, the actual `GetBlockX` calls are very slow. And since we must finish them all before committing to the DB, the progress is lost if the wallet is restarted mid-way.

Should we consider writing to the DB each time N blockstamps have been fetched? (N=1 or maybe bigger?)

Reply:
This is already called in batches of 10K from `BitcoindClient`, a ticker is used within `RPCClient` to send them periodically, and `NeutrinoClient` only sends a final `OnRescanFinished` notification. We could batch them here if it is an issue though.